### ISSUE TITLE
[Feature Request] Spotify Casting capabilities in the media browser.

### ISSUE THREAD

**Issue by rako77** [fondberg/spotcast#233]
I want to make a feature to allow Spotify to cast to Chromecast devices from the media browser. 

I'm willing to do the brunt of the dev work on this.
I think the best place for a feature like this is in the official Spotify integration but as an unofficial Spotify API like the token you need to cast to a Chromecast device is unlikely to make it into home assistant core, I think this plugin is the best place to do the work.

I am unfamilia with the media browser API but as this plugin can call the official plugin I hope this as easy registering spotcast as a media source for Chromecast devices and then just using a shim and have official Spotify plugin to show the media browser. I realise it might be a lot more complicated then this but I think I would like to give it a try anyway.

Firstly I would like your blessing that you would accept a PR that does this.

Secondly I would like to use this thread to discuss implementation details.

**Comment by fondberg** [comment]
Hi!

sounds interesting. Can you point to some docs on that API?

**Comment by fondberg** [comment]
@rako77 can you elaborate?

**Comment by skank01** [comment]
This would rock !

**Comment by skank01** [comment]
> Hi!
> 
> sounds interesting. Can you point to some docs on that API?

I dont think theres an api for media browser
Media browser is part of HA
Not sure if @rako77 is willing to elaborate, since theres no reply here

**Comment by skank01** [comment]
Since new people are joining further development here, is this on the road, i think many people wants this.
I find it still strange this hasnt been on developemnt ages ago (no offence meaning you guys dont progress hard enough.
All respect in what you guys are doing.
Take your needed time !

**Comment by gmcmicken** [comment]
@rako77 You could probably prototype this using "[universal media player](https://www.home-assistant.io/integrations/universal/)" and an input_select helper.

**Comment by Darkfull-Dante** [comment]
@rako77, @gmcmicken 

I was thinking about this request and i think a good place to start would be the xbox integration. It does have a media source browser [xbox integration dir](https://github.com/home-assistant/core/tree/dev/homeassistant/components/xbox).

Home assistant does have a library in media_player that can be accessed with
```python
from homeassistant.components.media_player import BrowseMedia
```

I'll start exploreing the possibilities. I think we must try to get our hand on the documentation for BrowseMedia first

**Comment by emontnemery** [comment]
The official Spotify integration now has media browsing support: https://github.com/home-assistant/core/pull/64921

This is essentially what's needed in the cast media_player to add support there: https://github.com/home-assistant/core/commit/deb196af36e3bba012a5becca764e2ac6bda7252

Would it make sense to allow spotcast to some hooks in the cast media player to enable spotify support there?

**Comment by gmcmicken** [comment]
> thank you @emontnemery, good thing that it was added upstream. I was working on a proof of concept on my side to add it to spotcast, but will probably be simpler that way.
> 
> I'm going to read on the documentation and check how the media player request could be sent to Spotcast instead of directly to the media_player itself.

Like I mentioned, I think you can combine two media players with the universal media player component and possibly do this right out of the box already.

**Comment by emontnemery** [comment]
Media browsing as well as support for the cast media player entity to play spotify content with help of spotcast is implemented by #294

**Comment by emontnemery** [comment]
@adrianmihalko You need Home Assistant Core 2022.2, and both spotify and spotcast integrations setup.

**Comment by fcusson** [comment]
@adrianmihalko , the cast module only permits you to browse media from a specific device. You can check #294 which documents how to access it (you need to open the media browser from the media player directly). We would need to implement a media_source browser, but the spotify integration doesn't seem to permit that for the moment. We are still evaluating the feasibility of that part.

**Comment by emontnemery** [comment]
@fcusson Oh, now I think I finally understand what you mean, you want it to be possible to browse the Spotify library also directly via the "Media" view in the frontend?

This is already possible, we will populate the root depending on which player is selected 👍 
Note that the screenshots below is with latest dev version of Home Assistant + #294.

This is with the default in-browser player selected, only local media + cameras can be browsed:
![image](https://user-images.githubusercontent.com/14281572/153609009-fe7e96fb-6f8a-4c74-9da8-5702d2325cda.png)

This is with the spotify player selected, only spotify can be browsed:
![image](https://user-images.githubusercontent.com/14281572/153609179-1ba39b9e-e35f-452c-87a3-a0e86280c0e3.png)

This is with a Google Chromecast Audio player selected, everything can be browsed:
![image](https://user-images.githubusercontent.com/14281572/153609316-df917dd8-fa69-406e-9691-ca76010dfe45.png)

**Comment by fcusson** [comment]
oh wow, thank you @emontnemery I really missed that part completely. So, I'm just waiting for fondberg to provide its review and we'll be able to merge the PR. There will be a 3.6.25 release but that will be a bug fix release to ensure compatibility with python 3.8 for certain users.

**Comment by adrianmihalko** [comment]
I can confirm, it works great now.

<img width="1027" alt="Screenshot 2022-02-11 at 16 48 45" src="https://user-images.githubusercontent.com/733361/153623156-a6c124db-035f-4ade-a0e4-c0ad966869f2.png">

The only thing I didn't really like in this Media player that seek is not working here (jump on the timeline where I want)?

**Comment by skank01** [comment]
> @fcusson Oh, now I think I finally understand what you mean, you want it to be possible to browse the Spotify library also directly via the "Media" view in the frontend?
> 
> This is already possible, Home Assistant populates the media browser root depending on which player is selected 👍 Home Assistant will even add a root node for each Spotify account once this PR is merged: [home-assistant/core#66256](https://github.com/home-assistant/core/pull/66256)
> 
> Note that the screenshots below is with latest dev version of Home Assistant + #294.
> 
> This is with the default in-browser player selected, only local media + cameras can be browsed: ![image](https://user-images.githubusercontent.com/14281572/153609009-fe7e96fb-6f8a-4c74-9da8-5702d2325cda.png)
> 
> This is with the spotify player selected, only spotify can be browsed: ![image](https://user-images.githubusercontent.com/14281572/153609179-1ba39b9e-e35f-452c-87a3-a0e86280c0e3.png)
> 
> This is with a Google Chromecast Audio player selected, everything can be browsed: ![image](https://user-images.githubusercontent.com/14281572/153609316-df917dd8-fa69-406e-9691-ca76010dfe45.png)

Huh how do you do that?
I can only browse my selected spotify media player, and then browse that media.
If i select a cast device, it says de media is empty

What i want to be able to do is: select a mediaplayer; then chose the spotify account and then browse that media... but i cant

Also... the cameras aren there either

**Comment by fcusson** [comment]
Hi @skank01, to get the casting ability, you would need to manually add a file `cast.py`. This file is currently under review in #294 and should be available as a release soon.

For your other questions:

- selecting a Spotify account is not currently possible, but will be in a near futur when the Spotify integration adds support for in [PR 66256]( https://github.com/home-assistant/core/pull/66256)
- cameras will only show if you have cameras setup and have a compatible media_player (which support video) selected
